datasource/src/main/scala/quasar/physical/blobstore/azure/AzureDatasource.scala

```diff
@@ -14,34 +14,88 @@
  * limitations under the License.
  */
 
-package quasar.physical.blobstore.azure
+package quasar.physical
+package blobstore
+package azure
 
+import slamdata.Predef._
 import quasar.api.datasource.DatasourceType
-import quasar.blobstore.ResourceType
-import quasar.blobstore.azure.{Azure, AzureBlobstore, AzureConfig, MaxQueueSize}
-import quasar.connector.MonadResourceErr
+import quasar.api.resource.{ResourceName, ResourcePath, ResourcePathType}
+import quasar.blobstore.{BlobstoreStatus, ResourceType}
+import quasar.blobstore.azure.{converters => azureConverters, _}
+import quasar.blobstore.paths.{BlobPath, BlobstorePath, PrefixPath}
+import quasar.blobstore.services.{GetService, ListService, PropsService}
+import quasar.connector.{MonadResourceErr, ResourceError}
 import quasar.connector.ParsableType.JsonVariant
-import quasar.physical.blobstore.BlobstoreDatasource
 
-import cats.Applicative
+import java.lang.Integer
+
+import cats.Monad
+import cats.data.Kleisli
 import cats.effect.ConcurrentEffect
+import cats.instances.int._
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import cats.syntax.eq._
 import cats.syntax.functor._
+import com.microsoft.azure.storage.blob.StorageException
 import eu.timepit.refined.auto._
-import fs2.RaiseThrowable
+import fs2.{RaiseThrowable, Stream}
 
-class AzureDatasource[F[_]: Applicative: MonadResourceErr: RaiseThrowable](
-  azureBlobstore: AzureBlobstore[F],
+class AzureDatasource[
+  F[_]: Monad: MonadResourceErr: RaiseThrowable, BP, PP](
+  resourcePathToBlobPath: Kleisli[F, ResourcePath, BP],
+  resourcePathToPrefixPath: Kleisli[F, ResourcePath, PP],
+  status: F[BlobstoreStatus],
+  prefixPathList: ListService[F, PP, (ResourceName, ResourcePathType)],
+  blobPathIsValid: PropsService[F, BP, Boolean],
+  blobPathGet: GetService[F, BP],
   jsonVariant: JsonVariant)
-  extends BlobstoreDatasource[F](AzureDatasource.dsType, jsonVariant, azureBlobstore)
+  extends BlobstoreDatasource[F, BP, PP](
+    AzureDatasource.dsType,
+    jsonVariant,
+    resourcePathToBlobPath,
+    resourcePathToPrefixPath,
+    status,
+    prefixPathList,
+    blobPathIsValid,
+    blobPathGet)
 
 object AzureDatasource {
   val dsType: DatasourceType = DatasourceType("azure", 1L)
 
-  def mk[F[_]: ConcurrentEffect: MonadResourceErr](cfg: AzureConfig): F[AzureDatasource[F]] =
-    Azure.mkContainerUrl[F](cfg)
-      .map(c => new AzureDatasource[F](
-        new AzureBlobstore(c, cfg.maxQueueSize.getOrElse(MaxQueueSize.default)),
-        toJsonVariant(cfg.resourceType)))
+  private def errorHandler[F[_]: RaiseThrowable, A](path: BlobPath): Throwable => Stream[F, A] = {
+    case ex: StorageException if ex.statusCode() === 404 =>
+      Stream.raiseError(ResourceError.throwableP(ResourceError.pathNotFound(converters.blobPathToResourcePath(path))))
```

Comment:
It would be preferable to raise this via `MonadResourceErr` instead of converting it to a throwable.